src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -433,23 +453,49 @@ private void PopulateReferenceIdDictionary(IEnumerable<EntryComponent> bundleEnt
         {
             foreach (EntryComponent entry in bundleEntries)
             {
-                if (entry.Request.Method != HTTPVerb.POST)
-                {
-                    continue;
-                }
+                var versionId = string.Empty;
 
-                // We've already come across this ID
-                if (!string.IsNullOrWhiteSpace(entry.FullUrl) && !idDictionary.ContainsKey(entry.FullUrl))
+                if (!string.IsNullOrWhiteSpace(entry.FullUrl))
                 {
-                    // This id is new to us
-                    var insertId = _resourceIdProvider.Create();
-                    entry.Resource.Id = insertId;
+                    if (entry.FullUrl.Contains("/_history/", StringComparison.OrdinalIgnoreCase))
+                    {
+                        string[] versionedReference = entry.FullUrl.Split("/_history/");
+                        versionId = versionedReference[1];
+                    }
 
-                    idDictionary.Add(entry.FullUrl, (insertId, entry.Resource.TypeName));
+                    var requestUrl = entry.Request?.Url;
+
+                    if (entry.Request.Method == HTTPVerb.POST)
+                    {
+                        // We've already come across this ID
+                        if (!idDictionary.ContainsKey(entry.FullUrl))
+                        {
+                            // This id is new to us
+                            var insertId = _resourceIdProvider.Create();
+                            entry.Resource.Id = insertId;
+
+                            idDictionary.Add(entry.FullUrl, (insertId, entry.Resource.TypeName));
+                        }
+                    }
+                    else if (!string.IsNullOrWhiteSpace(versionId) && requestUrl != null)
```

Comment:
Does it matter if this is a `GET` request?

Reply:
This code flow will be executed only if fullUrl is present, which means we will ignore get and delete calls.